x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,341 @@
+
+    private final ShardFollowTask params;
+    private final TimeValue retryTimeout;
+    private final TimeValue idleShardChangesRequestDelay;
+    private final BiConsumer<TimeValue, Runnable> scheduler;
 
-    private final AtomicLong processedGlobalCheckpoint = new AtomicLong();
+    private volatile long lastRequestedSeqno;
+    private volatile long globalCheckpoint;
 
-    ShardFollowNodeTask(long id, String type, String action, String description, TaskId parentTask, Map<String, String> headers) {
+    private volatile int numConcurrentReads = 0;
+    private volatile int numConcurrentWrites = 0;
+    private volatile long processedGlobalCheckpoint = 0;
+    private volatile long currentIndexMetadataVersion = 0;
+    private final AtomicInteger retryCounter = new AtomicInteger(0);
+    private final Queue<Translog.Operation> buffer = new PriorityQueue<>(Comparator.comparing(Translog.Operation::seqNo).reversed());
+
+    ShardFollowNodeTask(long id, String type, String action, String description, TaskId parentTask, Map<String, String> headers,
+                        ShardFollowTask params, BiConsumer<TimeValue, Runnable> scheduler, TimeValue idleShardChangesRequestDelay,
+                        TimeValue retryTimeout) {
         super(id, type, action, description, parentTask, headers);
+        this.params = params;
+        this.scheduler = scheduler;
+        this.retryTimeout = retryTimeout;
+        this.idleShardChangesRequestDelay = idleShardChangesRequestDelay;
+    }
+
+    void start(long followGlobalCheckpoint) {
+        this.lastRequestedSeqno = followGlobalCheckpoint;
+        this.processedGlobalCheckpoint = followGlobalCheckpoint;
+        this.globalCheckpoint = followGlobalCheckpoint;
+
+        // Forcefully updates follower mapping, this gets us the leader imd version and
+        // makes sure that leader and follower mapping are identical.
+        updateMapping(imdVersion -> {
+            currentIndexMetadataVersion = imdVersion;
+            LOGGER.info("{} Started to follow leader shard {}, followGlobalCheckPoint={}, indexMetaDataVersion={}",
+                params.getFollowShardId(), params.getLeaderShardId(), followGlobalCheckpoint, imdVersion);
+            coordinateReads();
+        });
     }
 
+    private synchronized void coordinateReads() {
+        if (isStopped()) {
+            LOGGER.info("{} shard follow task has been stopped", params.getFollowShardId());
+            return;
+        }
+
+        LOGGER.trace("{} coordinate reads, lastRequestedSeqno={}, globalCheckpoint={}",
+            params.getFollowShardId(), lastRequestedSeqno, globalCheckpoint);
+        final int maxReadSize = params.getMaxBatchOperationCount();
+        while (hasReadBudget() && lastRequestedSeqno < globalCheckpoint) {
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] read [{}/{}]", params.getFollowShardId(), numConcurrentReads, from, maxReadSize);
```

Comment:
nit: move this after the `maxRequiredSeqno` and use `maxRequiredSeqno` for the range?